sdk/nodejs/query/interfaces.ts

```diff
@@ -0,0 +1,44 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//
+// NOTE: We don't need `Disposable` or `Enumerator#reset()` yet, but it's worth noting here that we
+// didn't include them, and this causes us to diverge slightly from the "canonical" `Enumerable`
+// model.
+//
+
+export interface Enumerable<T> {
```

Comment:
Instead of defining our own `Enumerable`/`Enumerator` interfaces, why not build on ECMAScript 2015's `Iterable`/`Iterator` protocols?

 - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols
 - https://github.com/Microsoft/TypeScript/blob/342616781e80202f40872f044605806ba73837a3/src/lib/es2015.iterable.d.ts

```typescript
interface IteratorResult<T> {
    done: boolean;
    value: T;
}

interface Iterator<T> {
    next(value?: any): IteratorResult<T>;
    return?(value?: any): IteratorResult<T>;
    throw?(e?: any): IteratorResult<T>;
}

interface Iterable<T> {
    [Symbol.iterator](): Iterator<T>;
}
```

Then we could define our own `QueryableIterable<T>` and `QueryableIterablePromise<T>` interfaces that extend these to hang the query functions on.

Reply:
@justinvp I'm open to implementing it using iterators, but the truth is that when I read the docs I came away with no confidence that I understood how they're implemented, or what the technical implications of betting this on them would be. Since `Enumerable` is such a small part of the implementation here, I decided to just write it myself.

Some example questions I'm still confused about are: Does the error channel behave the same way as LINQ's? The completion channel? How hard is tree-shaking with iterators? Is there any advantage to using these protocols other than just being able to use `for`?

Reply:
@hausdorff These are good questions, that we should try to answer. My high-level feedback is that we should consider building on top of what the language provides, so that we can be as idiomatic as possible in a given language and fit-in well with its ecosystem. (And even potentially aligning with future language proposals, much in the same way that TypeScript does).

> Is there any advantage to using these protocols other than just being able to use `for`?

In addition to `for-of`, I believe it also enables destructuring and the spread operator (...) for iterables. There are proposals that expand on iterators that appear to very much align with our needs:

 - [Asynchronous Iterators for JavaScript](https://github.com/tc39/proposal-async-iteration)
 - [Iterator Helpers](https://github.com/tc39/proposal-iterator-helpers)

Implementing using iterators and/or async iterators would enable all sorts of third party "iterator helpers" to _just work_ with what we provide, without us having to provide an implementation or adapters ourselves.

Reply:
@justinvp do you think it's necessary to build the entire abstraction on top of iterators, or is it enough for our purposes to expose `toIterator()` and `fromIterator()` functions?

Reply:
@hausdorff, I think we should consider building the entire abstraction on top of iterators, unless there are good reasons not to beyond initial confusion/questions about how to implement them. If there are good reasons not to, I'd love to hear them, in which case, holding off and potentially providing `toIterator()` and `fromIterator()` adapters, if there's a need, sounds reasonable. Is there anything I can do to help clarify the initial uncertainty?

Also, I assume we eventually want to expose this in Python, right? Would we build on top of Python iterators so that things like `itertools` would _just work_ with our APIs, or are we going to build-out our own query operators on top of our own abstraction? (Apologies if this has already been discussed in a design meeting that I missed).

Reply:
@justinvp ok, we've started the move to an iterator-based combinator approach, where every query operator (_e.g._, `map`, `filter`) is "just" an async iterator.